### PR TITLE
Feature/native functions

### PR DESCRIPTION
**Pull request by collindutter**
----
📚 Documentation preview 📚: https://griptape--867.org.readthedocs.build//867/

Closes https://github.com/griptape-ai/griptape/issues/91

### CODE REVIEW

File: griptape/drivers/prompt/google_prompt_driver.py [hunk 1]
```diff
@@ -37,112 +47,198 @@ class GooglePromptDriver(BasePromptDriver):
...
+        protos = import_optional_dependency("google.generativeai.protos")
 
         if isinstance(content, TextMessageContent):
             return content.artifact.to_text()
         elif isinstance(content, ImageMessageContent):
             return ContentDict(mime_type=content.artifact.mime_type, data=content.artifact.value)
+        elif isinstance(content, ActionCallMessageContent):
+            action = content.artifact.value
+
+            return protos.Part(function_call=protos.FunctionCall(name=action.tag, args=action.input))
```
Comment (dylanholmes) [inline griptape/drivers/prompt/google_prompt_driver.py:201]: As I understand it the tool call (`function_call`) should only ever be part of messages with the role "model" / assistant.  Is there anything validating / guaranteeing this?

(Similar comment for the other relevant prompt drivers)

File: griptape/drivers/prompt/anthropic_prompt_driver.py [hunk 2]
```diff
@@ -87,44 +99,121 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
+        elif isinstance(artifact, (TextArtifact, ErrorArtifact, InfoArtifact)):
+            return {"type": "text", "text": artifact.to_text()}
+        else:
+            raise ValueError(f"Unsupported tool result artifact type: {type(artifact)}")
+
+    def __to_prompt_stack_message_content(self, content: ContentBlock) -> BaseMessageContent:
         if content.type == "text":
             return TextMessageContent(TextArtifact(content.text))
+        elif content.type == "tool_use":
+            name, path = content.name.split("_", 1)
```
Comment (dylanholmes) [inline griptape/drivers/prompt/anthropic_prompt_driver.py:185]: Is it possible for `action.name` to contain an underscore?

**Review by dylanholmes — COMMENTED** [review]
Still reviewing, but here's some initial comments

File: griptape/drivers/prompt/google_prompt_driver.py (hunk 1 above)
Comment (collindutter) [reply griptape/drivers/prompt/google_prompt_driver.py:201]: Yeah we don't really have any client-side validation on the Prompt Drivers to ensure the payload is formatted correctly. While I agree that this would be nice to have, it feels out of scope for this PR.

File: griptape/drivers/prompt/cohere_prompt_driver.py [hunk 3]
```diff
@@ -87,19 +99,145 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
+            ),
             **({"preamble": preamble} if preamble else {}),
         }
 
-    def __prompt_stack_content_message_content(self, content: BaseMessageContent) -> dict:
+    def __to_cohere_messages(self, messages: list[Message]) -> list[dict]:
+        cohere_messages = []
+
+        for message in messages:
+            new_message: dict = {"role": self.__to_cohere_role(message)}
```
Comment (dylanholmes) [inline griptape/drivers/prompt/cohere_prompt_driver.py:115]: Maybe rename `new_message` to `cohere_message` for readability

File: griptape/drivers/prompt/anthropic_prompt_driver.py (hunk 2 above)
Comment (collindutter) [reply griptape/drivers/prompt/anthropic_prompt_driver.py:185]: Unfortunately, yes. But the LLM providers are very particular about what characters can be included in the function name, and none of them support nested routes like we do (name/path).

The only thing protecting us here is that Actions are always Tools (right now at least), which are classes that _should_ follow python's class [naming convention](https://peps.python.org/pep-0008/#class-names). But this is a weak protection.

@vasinov we said we'd figure this out on the PR, so let's use this thread.

File: griptape/drivers/prompt/cohere_prompt_driver.py (hunk 3 above)
Comment (collindutter) [reply griptape/drivers/prompt/cohere_prompt_driver.py]: Done

File: griptape/drivers/prompt/cohere_prompt_driver.py [hunk 4]
```diff
@@ -87,19 +99,145 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
+
+            if message.has_any_content_type(ActionResultMessageContent):
+                new_message["tool_results"] = [
+                    self.__to_cohere_message_content(action_call)
+                    for action_call in message.get_content_type(ActionResultMessageContent)
```
Comment (dylanholmes) [inline griptape/drivers/prompt/cohere_prompt_driver.py:120]: Rename `action_call` to `action_result`

File: griptape/drivers/prompt/cohere_prompt_driver.py [hunk 5]
```diff
@@ -87,19 +99,145 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
+            action = content.artifact.value
+
+            return {"name": f"{action.name}_{action.path}", "parameters": action.input}
+        elif isinstance(content, ActionResultMessageContent):
+            artifact = content.artifact
+
+            if isinstance(artifact, ListArtifact):
+                message_content = [{"text": artifact.to_text()} for artifact in artifact.value]
+            else:
+                message_content = [{"text": artifact.to_text()}]
```
Comment (dylanholmes) [inline griptape/drivers/prompt/cohere_prompt_driver.py:147]: Do we know this will always be a `TextArtifact` or `ListArtifact`?

File: griptape/drivers/prompt/cohere_prompt_driver.py [hunk 6]
```diff
@@ -87,19 +99,145 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
+
+            return {
+                "call": {"name": f"{content.action.name}_{content.action.path}", "parameters": content.action.input},
```
Comment (dylanholmes) [inline griptape/drivers/prompt/cohere_prompt_driver.py:150]: Translation to and from  "{name}_{path}" happens in a few different files, might be nice to have a util to help keep consistent (maintainability/readability)

File: griptape/drivers/prompt/cohere_prompt_driver.py (hunk 5 above)
Comment (collindutter) [reply griptape/drivers/prompt/cohere_prompt_driver.py:147]: It's either a `ListArtifact` or some other Artifact that we're going to convert to text as best as we can.

File: griptape/drivers/prompt/cohere_prompt_driver.py (hunk 6 above)
Comment (collindutter) [reply griptape/drivers/prompt/cohere_prompt_driver.py:150]: Agreed, though let's fully agree on https://github.com/griptape-ai/griptape/pull/867#discussion_r1668658159 first.

File: griptape/drivers/prompt/anthropic_prompt_driver.py (hunk 2 above)
Comment (dylanholmes) [reply griptape/drivers/prompt/anthropic_prompt_driver.py:185]: I think that is a reasonable convention, so long as we call out somewhere in the Action/Tool docs that class names must not contain an underscore in order to properly work with native functions. Actually, you could ensure correctness by raising an exception if `action.name` contains an underscore.

That said, there are a couple places where split is called without the maxsplits arg. We should update those so that we don't need the same restriction on `path`.

File: griptape/drivers/prompt/cohere_prompt_driver.py [hunk 7]
```diff
@@ -87,19 +99,145 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
+            content.append(TextMessageContent(TextArtifact(response.text)))
+        if response.tool_calls is not None:
+            content.extend(
+                [
+                    ActionCallMessageContent(
+                        ActionArtifact(
+                            ActionArtifact.Action(
+                                tag=tool_call.name,
+                                name=tool_call.name.split("_")[0],
+                                path=tool_call.name.split("_")[1],
```
Comment (dylanholmes) [inline griptape/drivers/prompt/cohere_prompt_driver.py:216]: missing max split arg

File: griptape/drivers/prompt/google_prompt_driver.py [hunk 8]
```diff
@@ -37,112 +47,198 @@ class GooglePromptDriver(BasePromptDriver):
...
+        system_messages = prompt_stack.system_messages
+        if system_messages:
+            self.model_client._system_instruction = ContentDict(
+                role="system", parts=[Part(text=system_message.to_text()) for system_message in system_messages]
+            )
+
+        return {
+            "generation_config": GenerationConfig(
+                **{
+                    # For some reason, providing stop sequences when streaming breaks native functions
```
Comment (dylanholmes) [inline griptape/drivers/prompt/google_prompt_driver.py:118]: What's the impact on the output? Should we link or open a support case with google?

File: griptape/drivers/prompt/openai_chat_prompt_driver.py [hunk 9]
```diff
@@ -131,47 +139,152 @@ def _base_params(self, prompt_stack: PromptStack) -> dict:
...
         return params
 
-    def __to_role(self, message: Message) -> str:
+    def __to_openai_messages(self, messages: list[Message]) -> list[dict]:
+        openai_messages = []
+
+        for message in messages:
+            if message.is_text():
+                openai_messages.append({"role": message.role, "content": message.to_text()})
+            elif message.has_any_content_type(ActionResultMessageContent):
```
Comment (dylanholmes) [inline griptape/drivers/prompt/openai_chat_prompt_driver.py:154]: Can a message ever have a mix of `ActionResultMessageContent` and other message content types?

File: griptape/artifacts/action_artifact.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+from griptape.artifacts import BaseArtifact
+from griptape.mixins import SerializableMixin
+
+if TYPE_CHECKING:
+    from griptape.tools import BaseTool
+
+
+@define()
+class ActionArtifact(BaseArtifact, SerializableMixin):
```
Comment (dylanholmes) [inline griptape/artifacts/action_artifact.py:16]: After adding the "MessageContent" classes, do we still need an `ActionArtifact` class? Can you put an `Action` directly onto the MessageContent classes now?

(Raising as per offline discussion)

File: griptape/common/prompt_stack/prompt_stack.py [hunk 11]
```diff
@@ -1,14 +1,28 @@
...
-from griptape.common import Message, TextMessageContent, BaseMessageContent, ImageMessageContent
+
+if TYPE_CHECKING:
+    from griptape.tools import BaseTool
 
 
 @define
 class PromptStack(SerializableMixin):
     messages: list[Message] = field(factory=list, kw_only=True, metadata={"serializable": True})
+    actions: list[BaseTool] = field(factory=list, kw_only=True)
```
Comment (dylanholmes) [inline griptape/common/prompt_stack/prompt_stack.py:25]: Should this be "tools"? Since we have an `Action` class?

Comment (dylanholmes) [reply griptape/common/prompt_stack/prompt_stack.py:25]: As I understand it, actions are usages of tools. Is that correct?

File: griptape/drivers/prompt/cohere_prompt_driver.py (hunk 6 above)
Comment (collindutter) [reply griptape/drivers/prompt/cohere_prompt_driver.py]: Created utility in `Action`

File: griptape/drivers/prompt/anthropic_prompt_driver.py (hunk 2 above)
Comment (collindutter) [reply griptape/drivers/prompt/anthropic_prompt_driver.py]: I replaced all instances of `split` with a utility so there should be more consistency.

Will update the docs to call out the naming.

File: griptape/drivers/prompt/google_prompt_driver.py (hunk 8 above)
Comment (collindutter) [reply griptape/drivers/prompt/google_prompt_driver.py:119]: Created and linked to [an issue](https://github.com/google-gemini/generative-ai-python/issues/446)

File: griptape/drivers/prompt/openai_chat_prompt_driver.py (hunk 9 above)
Comment (collindutter) [reply griptape/drivers/prompt/openai_chat_prompt_driver.py:155]: As of right now, OpenAi exclusively returns function results -- so no there will not be a mix.

I'm not sure I follow your second question though.

File: griptape/artifacts/action_artifact.py (hunk 10 above)
Comment (collindutter) [reply griptape/artifacts/action_artifact.py:14]: Probably not. But I think an argument can be made that since an LLM can generate Actions, it makes sense as an Artifact. Also, splitting the Message Content types into "Artifact Message Content" "Non Artifact Message Content"  would be a very tedious refactor at this stage in the PR 😅

File: griptape/drivers/prompt/openai_chat_prompt_driver.py (hunk 9 above)
Comment (dylanholmes) [reply griptape/drivers/prompt/openai_chat_prompt_driver.py:155]: That's fine, this was not a very important comment/question

File: CHANGELOG.md [hunk 12]
```diff
@@ -55,6 +55,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Support for `PromptTask`s to take `TextArtifact`s, `ImageArtifact`s, and `ListArtifact`s as input.
 - Parameters `sort_key` and `sort_key_value` on `AmazonDynamoDbConversationMemoryDriver` for tables with sort keys.
 - `Reference` for supporting artifact citations in loaders and RAG engine modules.
+- Native function calling support to `OpenAiChatPromptDriver`, `AzureOpenAiChatPromptDriver`, `AnthropicPromptDriver`, `AmazonBedrockPromptDriver`, `GooglePromptDriver`, and `CoherePromptDriver`.
```
Comment (dylanholmes) [inline CHANGELOG.md:58]: 🎉

**Review by dylanholmes — DISMISSED** [review]
# ⚒️ ❗

**Review by dylanholmes — APPROVED** [review]

File: griptape/artifacts/action_artifact.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    """Represents an instance of an LLM calling a Action.
+
+    Attributes:
+        tag: The tag (unique identifier) of the action.
+        name: The name (Tool name) of the action.
+        path: The path (Tool activity name) of the action.
+        input: The input (Tool params) of the action.
+        tool: The matched Tool of the action.
+        output: The output (Tool result) of the action.
+    """
```
Comment (vasinov) [inline griptape/artifacts/action_artifact.py:24]: This docstring should be in `Action`.

File: griptape/common/action.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+from griptape.artifacts import BaseArtifact
+from griptape.mixins import SerializableMixin
+
+if TYPE_CHECKING:
+    from griptape.tools import BaseTool
+
+
+@define(kw_only=True)
+class Action(SerializableMixin):
```
Comment (vasinov) [inline griptape/common/action.py:15]: Should we add some prefix to this? I'm worried that we might have some other "action" in the future, unrelated to LLM actions, which will be confusing naming-wise.

File: griptape/common/action.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    name: str = field(metadata={"serializable": True})
+    path: Optional[str] = field(default=None, metadata={"serializable": True})
+    input: dict = field(factory=dict, metadata={"serializable": True})
+    tool: Optional[BaseTool] = field(default=None)
+    output: Optional[BaseArtifact] = field(default=None, metadata={"serializable": True})
+
+    def __str__(self) -> str:
+        return json.dumps(self.to_dict())
+
+    def to_dict(self) -> dict:
```
Comment (vasinov) [inline griptape/common/action.py:26]: Why did you need to override this? Can't we just set `tool` and `output` to non-serializable?

File: griptape/common/action.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        return {"tag": self.tag, "name": self.name, "path": self.path, "input": self.input}
+
+    def to_native_tool_name(self) -> str:
```
Comment (vasinov) [inline griptape/common/action.py:29]: I think we can drop "native" here since "tool" is a singular concept in the framework.

Comment (vasinov) [reply griptape/common/action.py:29]: Ah, I see that it makes sense to add a qualifier because of how we reference it in prompt drivers below. Should we call it "griptape_tool" for complete clarity?

File: griptape/common/action.py (hunk 14 above)
Comment (collindutter) [reply griptape/common/action.py:15]: `ToolAction`?

File: griptape/common/action.py (hunk 15 above)
Comment (collindutter) [reply griptape/common/action.py:26]: Yep, updated.

File: griptape/common/action.py (hunk 14 above)
Comment (collindutter) [reply griptape/common/action.py]: Refactored into `BaseAction` and `ToolAction`

File: griptape/artifacts/action_artifact.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+from griptape.artifacts import BaseArtifact
+from griptape.mixins import SerializableMixin
+
+if TYPE_CHECKING:
+    from griptape.common import ToolAction
+
+
+@define()
+class ActionArtifact(BaseArtifact, SerializableMixin):
```
Comment (vasinov) [inline griptape/artifacts/action_artifact.py:14]: Should this be renamed to `ToolActionArtifact`?

File: griptape/common/prompt_stack/contents/action_call_delta_message_content.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
+from __future__ import annotations
+from attrs import define, field
+from typing import Optional
+
+from griptape.common import BaseDeltaMessageContent
+
+
+@define
+class ActionCallDeltaMessageContent(BaseDeltaMessageContent):
```
Comment (vasinov) [inline griptape/common/prompt_stack/contents/action_call_delta_message_content.py:9]: Same with all events...I don't like the long names but we should probably stay consistent.

**Review by vasinov — CHANGES_REQUESTED** [review]

**Review by vasinov — APPROVED** [review]
